internal/pkg/providers/azure/azure.go

```diff
@@ -0,0 +1,238 @@
+
+	return securityRulesClient, nil
+}
+
+func deleteRule(rulesClient network.SecurityRulesClient, a *Azure, resourceName string, ipDescription string) error {
+	futureT, err := rulesClient.Delete(context.TODO(), a.ResourceGroupName, resourceName, ipDescription)
+	if err != nil {
+		return err
+	}
+	err = futureT.WaitForCompletionRef(context.TODO(), rulesClient.Client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getResourcesClient(a *Azure) (resources.Client, error) {
+
+	authorizer, err := auth.NewClientCredentialsConfig(a.ClientID, a.ClientSecret, a.TenantID).Authorizer()
+	if err != nil {
+		return resources.Client{}, err
+	}
+	resourcesClient := resources.NewClient(a.SubscriptionID)
+	resourcesClient.Authorizer = authorizer
+
+	return resourcesClient, nil
+}
+
+func getSecurityGroupsClient(a *Azure) (*network.SecurityGroupsClient, error) {
+
+	authorizer, err := auth.NewClientCredentialsConfig(a.ClientID, a.ClientSecret, a.TenantID).Authorizer()
+	if err != nil {
+		return nil, err
+	}
+	nsgClient := network.NewSecurityGroupsClient(a.SubscriptionID)
+	nsgClient.Authorizer = authorizer
+
+	return &nsgClient, nil
+}
+
+func createSecurityRule(a *Azure, rulesClient network.SecurityRulesClient, resourceName string, ipDescription *string, fromPortStr string, toPortStr string, ipRange *string, ipProtocol string) error {
+
+	var networkProtocol network.SecurityRuleProtocol
+
+	switch ipProtocol {
+	case "tcp":
+		networkProtocol = network.SecurityRuleProtocolTCP
+	case "udp":
+		networkProtocol = network.SecurityRuleProtocolUDP
+	case "icmp":
+		networkProtocol = network.SecurityRuleProtocolIcmp
+	case "esp":
+		networkProtocol = network.SecurityRuleProtocolEsp
+	case "*":
+		networkProtocol = network.SecurityRuleProtocolAsterisk
+	case "ah":
+		networkProtocol = network.SecurityRuleProtocolAh
+	default:
+		return errors.New("ip protocol " + ipProtocol + " unidentified")
+	}
+
+	future, err := rulesClient.CreateOrUpdate(
+		context.TODO(),
+		a.ResourceGroupName,
+		resourceName,
+		*ipDescription,
+		network.SecurityRule{
+			SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+				Access:                   network.SecurityRuleAccessAllow,
+				DestinationAddressPrefix: to.StringPtr("*"),
+				DestinationPortRange:     to.StringPtr(fromPortStr + "-" + toPortStr),
+				Direction:                network.SecurityRuleDirectionInbound,
+				Description:              ipDescription,
+				Priority:                 to.Int32Ptr(100),
+				Protocol:                 networkProtocol,
+				SourceAddressPrefix:      ipRange,
+				SourcePortRange:          to.StringPtr("*"),
+			},
+		})
+	if err != nil {
+		return err
+	}
+	err = future.WaitForCompletionRef(context.TODO(), rulesClient.Client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func isSecurityRuleToBeRetained(existingRuleName string, ipPermissions []utils.IpPermission, keepRuleDescriptionPrefix string, removeRule bool) bool {
```

Comment:
This method should be replaced with two methods.
- One that verifies if this rule should be deleted based on the rules that user has provided i.e git repo.
- Second that verifies that the rule doesn't have the description to prevent it from being deleted.

Also these two methods should be wrapped in a third method which prevents these methods from being executed if `removeRule` is false.